sys/kern/imgact_elf.c

```diff
@@ -1680,17 +1681,22 @@ __elfN(puthdr)(struct thread *td, void *hdr, size_t hdrsize, int numsegs,
 #if defined(COMPAT_FREEBSD32) && __ELF_WORD_SIZE == 32
 	ehdr->e_machine = ELF_ARCH32;
 #else
-#ifdef CPU_CHERI
-	if (td->td_proc->p_sysent->sv_flags & SV_CHERI)
-		ehdr->e_machine = EM_MIPS_CHERI;
-	else
-#endif
 	ehdr->e_machine = ELF_ARCH;
 #endif
 	ehdr->e_version = EV_CURRENT;
 	ehdr->e_entry = 0;
 	ehdr->e_phoff = sizeof(Elf_Ehdr);
+#ifdef CPU_CHERI
+	if (td->td_proc->p_sysent->sv_flags & SV_CHERI) {
+		ehdr->e_flags |= EF_MIPS_ABI_CHERIABI;
+	}
+	if (CHERICAP_SIZE == 16)
+		ehdr->e_flags |= EF_MIPS_MACH_CHERI128;
+	else if (CHERICAP_SIZE == 32)
+		ehdr->e_flags |= EF_MIPS_MACH_CHERI256;
+#else
 	ehdr->e_flags = 0;
```

Comment:
Long term the fix here I think is to copy e_flags from the binary instead of trying to auto-generate a value.  We have a few issues on MIPS already due to having a zero e_flags.  I'll try to fix that in FreeBSD soon so we can pull it back into CheriBSD (copying e_flags from the binary).